library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/api/event/Event.java

```diff
@@ -0,0 +1,329 @@
+ * The factory methods for Event allows the user to provide an implementation of {@code T} which is used to
+ * execute the callbacks stored in this event instance. This allows a user to control how iteration works, whether an
+ * event is cancelled after a specific callback is executed or to make an event
+ * {@link ParameterInvokingEvent parameter invoking}.
+ * <p>
+ * Generally {@code T} should be a type which is a
+ * <a href="https://docs.oracle.com/javase/specs/jls/se16/html/jls-9.html#jls-9.8">functional interface</a>
+ * to allow callbacks registered to the event to be specified in a lambda, method reference form or implemented onto a
+ * class. A way to ensure that an interface is a functional interface is to place a {@link FunctionalInterface}
+ * annotation on the type. You can let T not be a functional interface, however it heavily complicates the process
+ * of implementing an invoker and only allows callback implementations to be done by implementing an interface onto a
+ * class or extending a class.
+ *
+ * <h2>Example: Registering callbacks</h2>
+ * <p>
+ * The most common use of an event will be registering a callback which is executed by the event. To register a callback,
+ * pass an instance of {@code T} into {@link #register}.
+ *
+ * <pre>{@code
+ * // Events should use a dedicated functional interface for T rather than overloading multiple events to the same type
+ * // to allow those who implement using a class to implement multiple events.
+ * @FunctionalInterface
+ * public interface Example {
+ *     void doSomething();
+ * }
+ *
+ * // You can also return this instance of Event from a method, may be useful where a parameter is needed to get
+ * // the right instance of Event.
+ * public static final Event<Example> EXAMPLE = Event.create(...); // implementation
+ *
+ * public void registerEvents() {
+ *     // Since T is a functional interface, we can use the lambda form.
+ *     EXAMPLE.register(() -> {
+ *         // Do something
+ *     });
+ *
+ *     // Or we can use a method reference.
+ *     EXAMPLE.register(this::runSomething);
+ *
+ *     // Or implement T using a class.
+ *     // You can also use an anonymous class here; for brevity that is not included.
+ *     EXAMPLE.register(new ImplementedOntoClass());
+ * }
+ *
+ * public void runSomething() {
+ *     // Do something else
+ * }
+ *
+ * // When implementing onto a class, the class must implement the same type as the event invoker.
+ * class ImplementedOntoClass implements Example {
+ *     public void doSomething() {
+ *         // Do something else again
+ *     }
+ * }
+ * }</pre>
+ *
+ * <h2>Example: Executing an event</h2>
+ * <p>
+ * Executing an event is done by calling a method on the event invoker. Where {@code T} is Example, executing an event
+ * is done through the following:
+ *
+ * <pre>{@code
+ * EXAMPLE.invoker().doSomething();
+ * }</pre>
+ *
+ * @param <T> the type of the invoker used to execute an event and the type of the callback
```

Comment:
We should mention how to use the phase system here.

Reply:
Added.